src/machine/usb/hcd_mimxrt1062.go

```diff
@@ -0,0 +1,130 @@
+const hcdInterruptPriority = 3
+
+// hostController implements USB host controller driver (hcd) interface.
+type hostController struct {
+	core *core // Parent USB core this instance is attached to
+	port int   // USB port index
+	cc   class // USB host class
+	id   int   // hostControllerInstance index
+
+	bus *nxp.USB_Type
+	phy *nxp.USBPHY_Type
+	irq interrupt.Interrupt
+
+	cri volatile.Register8 // set to 1 if in critical section, else 0
+	ivm uintptr            // interrupt state when entering critical section
+}
+
+// hostControllerInstance provides statically-allocated instances of each USB
+// host controller configured on this platform.
+var hostControllerInstance [hcdCount]hostController
+
+// initHCD initializes and assigns a free host controller instance to the given
+// USB port. Returns the initialized host controller or nil if no free host
+// controller instances remain.
+func initHCD(port int, class class) (hcd, status) {
+	if 0 == hcdCount {
+		return nil, statusInvalid // must have defined host controllers
+	}
+	// Return the first instance whose assigned core is currently nil.
+	for i := range hostControllerInstance {
+		if nil == hostControllerInstance[i].core {
+			// Initialize host controller.
+			hostControllerInstance[i].core = &coreInstance[port]
+			hostControllerInstance[i].port = port
+			hostControllerInstance[i].cc = class
+			hostControllerInstance[i].id = i
+			switch port {
+			case 0:
+				hostControllerInstance[i].bus = nxp.USB1
+				hostControllerInstance[i].phy = nxp.USBPHY1
+				hostControllerInstance[i].irq =
+					interrupt.New(nxp.IRQ_USB_OTG1,
+						func(interrupt.Interrupt) {
+							coreInstance[0].hc.interrupt()
+						})
+
+			case 1:
+				hostControllerInstance[i].bus = nxp.USB2
+				hostControllerInstance[i].phy = nxp.USBPHY2
+				hostControllerInstance[i].irq =
+					interrupt.New(nxp.IRQ_USB_OTG2,
+						func(interrupt.Interrupt) {
+							//coreInstance[1].hc.interrupt()
+						})
+			}
+			return &hostControllerInstance[i], statusOK
+		}
+	}
+	return nil, statusBusy // No free host controller instances available.
+}
+
+func (hc *hostController) class() class { return hc.cc }
+
+func (hc *hostController) init() status {
+
+	return statusOK
+}
+
+func (hc *hostController) enable(enable bool) status {
+
+	hc.irq.SetPriority(hcdInterruptPriority)
+	hc.irq.Enable()
+
+	return statusOK
+}
+
+func (hc *hostController) critical(enter bool) status {
+	if enter {
+		// check if critical section already locked
+		if hc.cri.Get() != 0 {
+			return statusRetry
+		}
+		// lock critical section
+		hc.cri.Set(1)
+		// disable interrupts, storing state in receiver
+		hc.ivm = arm.DisableInterrupts()
+	} else {
+		// ensure critical section is locked
+		if hc.cri.Get() != 0 {
+			// re-enable interrupts, using state stored in receiver
+			arm.EnableInterrupts(hc.ivm)
+			// unlock critical section
+			hc.cri.Set(0)
+		}
+	}
+	return statusOK
+}
```

Comment:
I don't really understand this code. What kind of safety is it trying to provide? Specifically, I don't see in which case it would return `statusRetry`.

It appears that this code is not called from anywhere, maybe it can be removed?

If you just want to have a critical section, the best way to do this (which also supports nesting!), is using the `runtime/interrupt` package (`interrupt.Disable()` and `interrupt.Restore(mask)`).

Reply:
Right, old code, this needs to be removed. It was being called twice and causing a lockup, and then I realized it was just over-complicating things. Replaced with direct calls to either `(Interrupt).Disable()` or `arm.DisableInterrupts()` (both are necessary).

Reply:
`(Interrupt).Disable()` would be quite preferable, to at some point be able to support Espressif and/or RISC-V as well.